### PR TITLE
Do not allow to increase collateral factor of open market.

### PR DESCRIPTION
**Pull request by vivekvpandya**
## Issue
https://app.clickup.com/t/2953d0u

## Description
For an open lending market increase of collateral factor can trigger loans to be liquidated. So such updates are not allowed.

## Checklist

~- [ ] I have updated the cargo docs to reflect changes made by this PR _(if applicable)_~
~- [ ] I have updated the `book/` to reflect changes made by this PR _(if applicable)_~

### CODE REVIEW

File: frame/lending/src/tests.rs [hunk 1]
```diff
@@ -300,255 +301,255 @@ fn can_update_market() {
...
-					alice_total_debt_with_interest
-				);
-			},
-			_ => panic!(
-				"Unexpected value for System::events(); found {:#?}",
-				system_events
-			),
-		}
-	});
+            System::set_block_number(1); // ensure block is non-zero
```
Comment (benluelo) [inline frame/lending/src/tests.rs:304]: Is this all just whitespace change?

**Review by benluelo — APPROVED** [review]

**Comment by mikolaichuk** [comment]
@vivekvpandya sorry, have commited changes instead of request

File: frame/lending/src/tests.rs (hunk 1 above)
Comment (vivekvpandya) [reply frame/lending/src/tests.rs:304]: I think it is done by cargo fmt

**Comment by vivekvpandya** [comment]
> @vivekvpandya sorry, have commited changes instead of request

Could you please explain it? To me updated check seems wrong, or I am missing some context.

**Comment by mikolaichuk** [comment]
> > @vivekvpandya sorry, have commited changes instead of request
> 
> Could you please explain it? To me updated check seems wrong, or I am missing some context.

@vivekvpandya we update 'market.collateral.factor' with 'input.collateral.factor'. Market collateral factor should not be increased, so if `input.collateral.factor > market.collateral.factor` we  should not update 'market.collateral.factor' with new value.

File: frame/lending/src/tests.rs [hunk 2]
```diff
@@ -1589,9 +1586,12 @@ fn market_owner_cannot_retroactively_liquidate() {
...
 		};
 		// ALICE is the creater of the market.
-		assert_ok!(Lending::update_market(Origin::signed(*ALICE), market_id, updatable));
-		// BOB loan must be liquidated now.
-		assert_eq!(Lending::should_liquidate(&market_id, &BOB), Ok(true));
+		assert_noop!(
+			Lending::update_market(Origin::signed(*ALICE), market_id, updatable),
+			Error::<Runtime>::CannotIncreaseCollateralFactorOfOpenMarket
+		);
+		// if above update was succeeded BOB's loan would have to be liquidated.
```
Comment (mikolaichuk) [inline frame/lending/src/tests.rs:1593]: Should we drop this accert?

**Review by mikolaichuk — APPROVED** [review]

File: frame/lending/src/tests.rs [hunk 3]
```diff
@@ -1589,9 +1586,12 @@ fn market_owner_cannot_retroactively_liquidate() {
...
+		assert_eq!(Lending::should_liquidate(&market_id, &BOB), Ok(false));
```
Comment (mikolaichuk) [inline frame/lending/src/tests.rs:1594]: Should we drop this assert?

**Review by mikolaichuk — APPROVED** [review]

File: frame/lending/src/tests.rs (hunk 3 above)
Comment (vivekvpandya) [reply frame/lending/src/tests.rs:1594]: @mikolaichuk ensure macro retruns error if its condition is not satisfied . So my original check was correct and your check is wrong
```rust
					ensure!(
						input.collateral_factor > market.collateral_factor,
						Error::<T>::CannotIncreaseCollateralFactorOfOpenMarket
					);
```
with above check function will only allow updates when input > market's collateral.

Comment (vivekvpandya) [reply frame/lending/src/tests.rs:1594]: Also after my change updates which increases collateral factor should not be allowed and hence following assert is correct.
```rust
assert_eq!(Lending::should_liquidate(&market_id, &BOB), Ok(false));
```

**Review by dzmitry-lahoda — COMMENTED** [review]
was it product requirement? if this is temporal solution before governance, can we have a comment?

**Review by mikolaichuk — DISMISSED** [review]

**Comment by dzmitry-lahoda** [comment]
> Market collateral factor should not be increased, so if input.collateral.factor > market.collateral.factor we should not update 'market.collateral.factor' with new value.

why not? and if has special treatment untlike other parameters which can change curve and any other curve parameter, may be separate extinsic?

**Comment by mikolaichuk** [comment]
> > Market collateral factor should not be increased, so if input.collateral.factor > market.collateral.factor we should not update 'market.collateral.factor' with new value.
> 
> why not? and if has special treatment untlike other parameters which can change curve and any other curve parameter, may be separate extinsic?

I believe  this question is addressed to @vivekvpandya.
I do not now about other parameters, but if market owner can change collateral factor after the market was started, he can initiate liquidation for all positions within the market by changing collateral factor to inadequate value. Is it expected behaviour?

**Comment by vivekvpandya** [comment]
> > > Market collateral factor should not be increased, so if input.collateral.factor > market.collateral.factor we should not update 'market.collateral.factor' with new value.
> > 
> > 
> > why not? and if has special treatment untlike other parameters which can change curve and any other curve parameter, may be separate extinsic?
> 
> I believe this question is addressed to @vivekvpandya. I do not know about other parameters, but if market owner can change collateral factor after the market was started, he can initiate liquidation for all positions within the market by changing collateral factor to inadequate value. Is it expected behaviour?
For now I am told to not allow increase in collateral factor. We can keep clickup ticket open for discussion more better solutions.

**Comment by dzmitry-lahoda** [comment]
>  but if market owner can change collateral factor after the market was started, he can initiate liquidation for all positions within the market by changing collateral factor to inadequate value. Is it expected behaviour?

yes. yes.

also he can change curve, increase IR curve parameters from 2 to 99 and hit liquidation.

**Comment by vivekvpandya** [comment]
> > but if market owner can change collateral factor after the market was started, he can initiate liquidation for all positions within the market by changing collateral factor to inadequate value. Is it expected behaviour?
> 
> yes. yes.
> 
> also he can change curve, increase IR curve parameters from 2 to 99 and hit liquidation.

I would suggest IR should not be up-gradable once market is open. If we agree it can be done in separate PR with test.

**Comment by mikolaichuk** [comment]
> > but if market owner can change collateral factor after the market was started, he can initiate liquidation for all positions within the market by changing collateral factor to inadequate value. Is it expected behaviour?
> 
> yes. yes.
> 
> also he can change curve, increase IR curve parameters from 2 to 99 and hit liquidation.

I would not borrow money on those terms :)

**Comment by dzmitry-lahoda** [comment]
> I would not borrow money on those terms :)


do you have 1M usd? :)

> I would suggest IR should not be up-gradable once market is open. If we agree it can be done in separate PR with test.

I suggest to get approve from PO and write doc updates into reame or mdbook about.

**Comment by mikolaichuk** [comment]
> do you have 1M usd? :)
Not yet :(

**Review by benluelo — APPROVED** [review]

**Review by mikolaichuk — APPROVED** [review]

**Comment by mikolaichuk** [comment]
bors merge

**Comment by mikolaichuk** [comment]
bors ping

**Pull request merged**